_We're going to be using pandas an json, so lets start with that._

In [1]:
import json
import pandas as pd
import os
from pandas.io.json import json_normalize
import geopandas

# Looking at the json files

_Before we do any real work, can we list the files from our target directory?_

In [2]:
!pwd
! ls ../APIs

/home/vagrant/git/BigData/democratizing_weather_data/democratizing_weather_data/streaming/prototypes
2yahoo.json	     weather.json	      yahoo_weather1line.json
API_index.csv	     wsdot.TrafficFlows.json  yahoo_weather2.json
openweathermap.json  wsdot.Traveltimes.json   yahoo_weather.json
weather2.json	     yahoo_weather


_Nice! Let's examine a json file_

In [3]:
!head ../APIs/yahoo_weather.json

{
    "query": {
        "count": 1,
        "created": "2017-07-29T23:51:25Z",
        "lang": "en-US",
        "diagnostics": {
            "publiclyCallable": "true",
            "url": {
                "execution-start-time": "1",
                "execution-stop-time": "5",


# Exploring JSON from the Yahoo Weather API

In [4]:
!ls wsdot_weather/traffic
!cat wsdot_weather/traffic/2017-08-13-23-09-39 

2017-08-13-23-09-39  2017-08-13-23-11-39  2017-08-13-23-39-28
2017-08-13-23-10-39  2017-08-13-23-12-39  2017-08-14-00-09-28


[{"AverageTime":31,"CurrentTime":34,"Description":"Everett to Downtown Seattle using HOV lanes","Distance":26.72,"EndPoint":{"Description":"I-5 @ University St in Seattle","Direction":"S","Latitude":47.609294000,"Longitude":-122.331759000,"MilePost":165.83,"RoadName":"I-5"},"Name":"HOV Everett-Seattle (SB AM)","StartPoint":{"Description":"I-5 @ 41st St in Everett","Direction":"S","Latitude":47.924280000,"Longitude":-122.265480000,"MilePost":192.55,"RoadName":"I-5"},"TimeUpdated":"\/Date(1502665500000-0700)\/","TravelTimeID":2},{"AverageTime":27,"CurrentTime":27,"Description":"Downtown Seattle to Everett using HOV lanes","Distance":26.94,"EndPoint":{"Description":"I-5 @ 41st St in Everett","Direction":"N","Latitude":47.924280000,"Longitude":-122.265480000,"MilePost":192.77,"RoadName":"I-5"},"Name":"HOV Seattle-Everett (NB PM)","StartPoint":{"Description":"I-5 @ University St in Seattle","Direction":"N","Latitude":47.609294000,"Longitude":-122.331759000,"MilePost":165.83,"RoadName":"I-5"

In [5]:
df = pd.read_json('wsdot_weather/traffic/2017-08-13-23-09-39')
df

AverageTime  CurrentTime  \
0             31           34   
1             27           27   
2             28           27   
3             11           12   
4             11           11   
5             10           10   
6             11           13   
7             26           26   
8             27           27   
9             26           26   
10            27           27   
11            18           31   
12            12           10   
13             9            9   
14             9            9   
15            34           46   
16            28           24   
17            22           22   
18            23           22   
19            27           35   
20            37           49   
21            27           27   
22            30           29   
23            16           16   
24            19           22   
25            17           16   
26             0            0   
27             0            0   
28            16           17   
29            14           14   
..           ...          ...   
150           25           25   
151           14           15   
152           13           14   
153           19           20   
154           14           15   
155           14           14   
156           45           18   
157            7            8   
158           15           15   
159           26           27   
160           23           24   
161           20           21   
162           35           37   
163           35           35   
164            8           12   
165           13           17   
166           10           10   
167           23           22   
168           17           18   
169           10           10   
170           21           21   
171           18           18   
172           10            9   
173            3            3   
174           17           17   
175            5            4   
176           11           10   
177           21           19   
178           10            7   
179           15           13   

                                           Description  Distance  \
0          Everett to Downtown Seattle using HOV lanes     26.72   
1          Downtown Seattle to Everett using HOV lanes     26.94   
2                          Downtown Seattle to Everett     26.94   
3                        Downtown Bellevue to Issaquah      9.55   
4        Downtown Bellevue to Issaquah using HOV lanes      9.55   
5        Issaquah to Downtown Bellevue using HOV lanes      9.48   
6                        Issaquah to Downtown Bellevue      9.48   
7                         Downtown Bellevue to Everett     26.06   
8                         Everett to Downtown Bellevue     26.34   
9         Downtown Bellevue to Everett using HOV lanes     26.06   
10        Everett to Downtown Bellevue using HOV lanes     26.34   
11                               Federal Way to SeaTac      8.82   
12               Federal Way to SeaTac using HOV lanes      8.82   
13               SeaTac to Federal Way using HOV lanes      9.16   
14                               SeaTac to Federal Way      9.16   
15                     Federal Way to Downtown Seattle     22.15   
16     Federal Way to Downtown Seattle using HOV lanes     22.15   
17     Downtown Seattle to Federal Way using HOV lanes     22.19   
18                     Downtown Seattle to Federal Way     22.19   
19                    Downtown Bellevue to Federal Way     24.56   
20                    Federal Way to Downtown Bellevue     23.58   
21    Downtown Bellevue to Federal Way using HOV lanes     24.56   
22    Federal Way to Downtown Bellevue using HOV lanes     23.58   
23        Downtown Seattle to Lynnwood using HOV lanes     15.73   
24                        Lynnwood to Downtown Seattle     15.23   
25                        Downtown Seattle to Lynnwood     15.73   
26   Lynnwood to Downtown Seattle using express lan...      0.00   
27    Lynnwood to Downtown Seattle using express lanes

In [6]:
!ls wsdot_weather/weather
!cat wsdot_weather/weather/2017-08-13-22-38-57 

2017-08-13-22-38-57  2017-08-13-22-42-57  2017-08-13-22-46-57
2017-08-13-22-39-57  2017-08-13-22-43-57  2017-08-13-22-47-57
2017-08-13-22-40-57  2017-08-13-22-44-57  2017-08-13-23-37-10
2017-08-13-22-41-57  2017-08-13-22-45-57  2017-08-14-00-07-10
[{"BarometricPressure":1010.60,"Latitude":47.474800000,"Longitude":-122.270400000,"PrecipitationInInches":null,"ReadingTime":"\/Date(1502663403000-0700)\/","RelativeHumidity":50,"SkyCoverage":"N\/A","StationID":1909,"StationName":"SB I-5 144th","TemperatureInFahrenheit":72.00,"Visibility":null,"WindDirection":235,"WindDirectionCardinal":"SW","WindGustSpeedInMPH":9,"WindSpeedInMPH":3},{"BarometricPressure":1014.80,"Latitude":47.760632547,"Longitude":-122.184047830,"PrecipitationInInches":null,"ReadingTime":"\/Date(1502663422000-0700)\/","RelativeHumidity":52,"SkyCoverage":"N\/A","StationID":1910,"StationName":"SB I-405 N.E. 195th","TemperatureInFahrenheit":72.00,"Visibility":null,"WindDirection":95,"WindDirectionCardinal":"E","WindGustSpeedInM

In [7]:
df = pd.read_json('wsdot_weather/weather/2017-08-13-22-38-57')
df

BarometricPressure   Latitude   Longitude  PrecipitationInInches  \
0                1010.6  47.474800 -122.270400                    NaN   
1                1014.8  47.760633 -122.184048                    NaN   
2                 981.5  47.509000 -121.885000                    NaN   
3                1002.9  47.726000 -122.324000                    NaN   
4                   NaN  46.436000 -117.350000                    NaN   
5                   NaN  47.033800 -123.097100                    NaN   
6                   NaN  47.555500 -120.586600                    NaN   
7                   NaN  47.269100 -121.239200                    NaN   
8                   NaN  48.070300 -124.049000                    NaN   
9                   NaN  46.887000 -120.427000                    NaN   
10                  NaN  46.928000 -120.500000                    NaN   
11                  NaN  47.305082 -121.305688                    NaN   
12                  NaN  47.078000 -120.724000                    NaN   
13                  NaN  46.943000 -120.165000                    NaN   
14                996.5  48.688000 -122.400000                    NaN   
15                  NaN  47.745900 -121.089100                    NaN   
16               1013.7  48.992000 -122.272000                    NaN   
17                  NaN  47.043000 -119.916000                    NaN   
18                  NaN  47.085000 -119.006000                    NaN   
19                  NaN  47.113000 -120.786000                    NaN   
20                  NaN  48.050400 -122.952300                    NaN   
21                  NaN  48.093800 -123.278300                    NaN   
22                  NaN  47.443500 -121.669400                    NaN   
23                995.1  47.473000 -121.742000                    NaN   
24                  NaN  47.427000 -121.418000                    NaN   
25                  NaN  47.334800 -120.577100                    NaN   
26                  NaN  46.638000 -121.390000                    NaN   
27                  NaN  47.301900 -120.067800                    NaN   
28                  NaN  47.233100 -119.958200                    NaN   
29                  NaN  47.815600 -119.363000                    NaN   
..                  ...        ...         ...                    ...   
73                  NaN  48.876000 -117.348300                    NaN   
74                998.8  47.532000 -121.985000                    NaN   
75                987.6  46.217600 -119.748900                    NaN   
76                976.8  46.538500 -120.471800                    NaN   
77                977.7  46.538400 -120.472000                    NaN   
78                  NaN  46.393000 -120.241100                    NaN   
79                981.9  46.326600 -120.077900                    NaN   
80                  NaN  47.067700 -122.715800                    NaN   
81                874.5  48.390400 -119.879200                    NaN   
82                  NaN  47.858400 -122.621600                    NaN   
83               1015.5  48.446000 -122.432000                    NaN   
84               1000.2  47.916300 -122.207200                    NaN   
85                926.8  47.980300 -122.188000                    NaN   
86               1012.0  48.204200 -122.214200                    NaN   
87                996.5  47.813900 -121.570900                    NaN   
88                882.0  46.500278 -119.611667                    NaN   
89                  NaN  46.467063 -119.541111                    NaN   
90                  NaN  46.383155 -119.422786                    NaN   
91                  NaN  46.325574 -119.316126                    NaN   
92               1011.2  47.303516 -122.232946                    NaN   
93                605.5  46.049483 -118.495917                    NaN   
94               1003.0  47.830967 -122.262359                    NaN   
95                  NaN  46.802261 -123.007164                 

_Read a json file from disk (we'll use ../API/<tbd>.json
We should reall use the yahoo api output, since we know how to convert that to a csv_

Source [Andy Hayden on ghthub.io](http://hayd.github.io/2013/pandas-json)

In [8]:
df = pd.read_json('../APIs/yahoo_weather.json')
df

query
count                                                        1
created                                   2017-07-29T23:51:25Z
diagnostics  {'publiclyCallable': 'true', 'url': {'executio...
lang                                                     en-US
results      {'channel': {'units': {'distance': 'mi', 'pres...

In [9]:
!cat ../APIs/yahoo_weather.json

{
    "query": {
        "count": 1,
        "created": "2017-07-29T23:51:25Z",
        "lang": "en-US",
        "diagnostics": {
            "publiclyCallable": "true",
            "url": {
                "execution-start-time": "1",
                "execution-stop-time": "5",
                "execution-time": "4",
                "content": "http://weather-ydn-yql.media.yahoo.com:4080/v3/public/weather/rss?w=2502265"
            },
            "user-time": "6",
            "service-time": "4",
            "build-version": "2.0.164"
        },
        "results": {
            "channel": {
                "units": {
                    "distance": "mi",
                    "pressure": "in",
                    "speed": "mph",
                    "temperature": "F"
                },
                "title": "Yahoo! Weather - Sunnyvale, CA, US",
                "link": "http://us.rd.yahoo.com/dailynews/rss/weather/Country__Country/*https://weather.yahoo.com/country/state/city-2502265/"

In [10]:
!cat ../APIs/weather2.json

{'clouds': {'all': 40}, 'name': 'Machias', 'visibility': 16093, 'sys': {'country': 'US', 'sunset': 1503026198, 'message': 0.0051, 'type': 1, 'id': 2931, 'sunrise': 1502975176}, 'weather': [{'main': 'Clouds', 'id': 802, 'icon': '03n', 'description': 'scattered clouds'}], 'coord': {'lat': 48, 'lon': -122}, 'base': 'stations', 'dt': 1502938380, 'main': {'pressure': 1019, 'temp_min': 290.15, 'temp_max': 298.15, 'temp': 294.1, 'humidity': 72}, 'id': 5802098, 'wind': {'speed': 3.6, 'deg': 330}, 'cod': 200}


In [11]:
#df = pd.read_json('../APIs/weather2.json')
#df

_Not much of a dataframe. Let's re-read and nomralize, drilling into the results section._

Read file with json Library. Source: [Scott Robinson's Stack Abuse Article ](http://stackabuse.com/reading-and-writing-json-to-a-file-in-python/)

In [12]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_df = json_normalize(raw_json['query']) 
yahoo_weather_df

count               created diagnostics.build-version  \
0      1  2017-07-29T23:51:25Z                   2.0.164   

  diagnostics.publiclyCallable diagnostics.service-time  \
0                         true                        4   

                             diagnostics.url.content  \
0  http://weather-ydn-yql.media.yahoo.com:4080/v3...   

  diagnostics.url.execution-start-time diagnostics.url.execution-stop-time  \
0                                    1                                   5   

  diagnostics.url.execution-time diagnostics.user-time  \
0                              4                     6   

             ...             results.channel.location.region  \
0            ...                                          CA   

                results.channel.title results.channel.ttl  \
0  Yahoo! Weather - Sunnyvale, CA, US                  60   

  results.channel.units.distance results.channel.units.pressure  \
0                             mi                             in   

  results.channel.units.speed results.channel.units.temperature  \
0                         mph                                 F   

  results.channel.wind.chill results.channel.wind.direction  \
0                         75                            355   

  results.channel.wind.speed  
0                         18  

[1 rows x 50 columns]

_One row, 50 columns. Let's look at data types._

In [13]:
yahoo_weather_df.dtypes

count                                     int64
created                                  object
diagnostics.build-version                object
diagnostics.publiclyCallable             object
diagnostics.service-time                 object
diagnostics.url.content                  object
diagnostics.url.execution-start-time     object
diagnostics.url.execution-stop-time      object
diagnostics.url.execution-time           object
diagnostics.user-time                    object
lang                                     object
results.channel.astronomy.sunrise        object
results.channel.astronomy.sunset         object
results.channel.atmosphere.humidity      object
results.channel.atmosphere.pressure      object
results.channel.atmosphere.rising        object
results.channel.atmosphere.visibility    object
results.channel.description              object
results.channel.image.height             object
results.channel.image.link               object
results.channel.image.title             

_That's a lot of data, all of type object. For the Yahoo Weather API, most of the good stuff lives in_ results.channel
_Let's dig deeper_

In [14]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_df = json_normalize(raw_json['query']['results']['channel']) 
yahoo_weather_df

astronomy.sunrise astronomy.sunset atmosphere.humidity atmosphere.pressure  \
0           6:10 am          8:19 pm                  58              1006.0   

  atmosphere.rising atmosphere.visibility  \
0                 0                  16.1   

                            description image.height  \
0  Yahoo! Weather for Sunnyvale, CA, US           18   

                 image.link     image.title    ...     location.region  \
0  http://weather.yahoo.com  Yahoo! Weather    ...                  CA   

                                title ttl units.distance units.pressure  \
0  Yahoo! Weather - Sunnyvale, CA, US  60             mi             in   

  units.speed units.temperature wind.chill wind.direction wind.speed  
0         mph                 F         75            355         18  

[1 rows x 39 columns]

_Looking more like a dataframe.  Let's look at fields & datatypes_

In [15]:
yahoo_weather_df.dtypes

astronomy.sunrise        object
astronomy.sunset         object
atmosphere.humidity      object
atmosphere.pressure      object
atmosphere.rising        object
atmosphere.visibility    object
description              object
image.height             object
image.link               object
image.title              object
image.url                object
image.width              object
item.condition.code      object
item.condition.date      object
item.condition.temp      object
item.condition.text      object
item.description         object
item.forecast            object
item.guid.isPermaLink    object
item.lat                 object
item.link                object
item.long                object
item.pubDate             object
item.title               object
language                 object
lastBuildDate            object
link                     object
location.city            object
location.country         object
location.region          object
title                    object
ttl     

_Everything's an object.  If we want to do any real analysis, we're going to need to do some type conversion. Also note, that the field names have a hierarchy to them.  So maybe we can drill deeper into just the fields we want_

_Drill Down: atmosphere_

In [16]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_atmosphere_df = json_normalize(raw_json['query']['results']['channel']['atmosphere']) 
yahoo_weather_atmosphere_df

humidity pressure rising visibility
0       58   1006.0      0       16.1

_Drill Down: image_

In [17]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_image_df = json_normalize(raw_json['query']['results']['channel']['image']) 
yahoo_weather_image_df

height                      link           title  \
0     18  http://weather.yahoo.com  Yahoo! Weather   

                                                 url width  
0  http://l.yimg.com/a/i/brand/purplelogo//uh/us/...   142

_Drill Down: item_

In [18]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_item_df = json_normalize(raw_json['query']['results']['channel']['item']) 
yahoo_weather_item_df

condition.code                 condition.date condition.temp condition.text  \
0             32  Sat, 29 Jul 2017 04:00 PM PDT             75          Sunny   

                                         description  \
0  <![CDATA[<img src="http://l.yimg.com/a/i/us/we...   

                                            forecast guid.isPermaLink  \
0  [{'code': '34', 'date': '29 Jul 2017', 'day': ...            false   

         lat                                               link         long  \
0  37.371609  http://us.rd.yahoo.com/dailynews/rss/weather/C...  -122.038254   

                         pubDate  \
0  Sat, 29 Jul 2017 04:00 PM PDT   

                                              title  
0  Conditions for Sunnyvale, CA, US at 04:00 PM PDT

_Drill Down: item.forecast_

In [19]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_forecast_df = json_normalize(raw_json['query']['results']['channel']['item']['forecast']) 
yahoo_weather_forecast_df

code         date  day high low           text
0   34  29 Jul 2017  Sat   78  58   Mostly Sunny
1   32  30 Jul 2017  Sun   84  60          Sunny
2   32  31 Jul 2017  Mon   87  61          Sunny
3   32  01 Aug 2017  Tue   88  64          Sunny
4   34  02 Aug 2017  Wed   90  66   Mostly Sunny
5   30  03 Aug 2017  Thu   88  68  Partly Cloudy
6   34  04 Aug 2017  Fri   88  67   Mostly Sunny
7   32  05 Aug 2017  Sat   86  66          Sunny
8   34  06 Aug 2017  Sun   83  63   Mostly Sunny
9   34  07 Aug 2017  Mon   79  61   Mostly Sunny

_Drill Down:condition_

In [20]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_conditions_df = json_normalize(raw_json['query']['results']['channel']['item']['condition']) 
yahoo_weather_conditions_df

code                           date temp   text
0   32  Sat, 29 Jul 2017 04:00 PM PDT   75  Sunny

In [21]:
yahoo_weather_conditions_df.dtypes

code    object
date    object
temp    object
text    object
dtype: object

_Drill Down: wind_

In [22]:
with open('../APIs/yahoo_weather.json') as json_file:
  raw_json = json.load(json_file)

yahoo_weather_atmosphere_df = json_normalize(raw_json['query']['results']['channel']['wind']) 
yahoo_weather_atmosphere_df

chill direction speed
0    75       355    18

# Exploring JSON from the Open Weather Map API

In [23]:
!pwd
! ls ../APIs
! cat ../APIs/openweathermap.json

/home/vagrant/git/BigData/democratizing_weather_data/democratizing_weather_data/streaming/prototypes
2yahoo.json	     weather.json	      yahoo_weather1line.json
API_index.csv	     wsdot.TrafficFlows.json  yahoo_weather2.json
openweathermap.json  wsdot.Traveltimes.json   yahoo_weather.json
weather2.json	     yahoo_weather
{
    "coord": {
        "lon": 139.01,
        "lat": 35.02
    },
    "weather": [
        {
            "id": 800,
            "main": "Clear",
            "description": "clear sky",
            "icon": "01n"
        }
    ],
    "base": "stations",
    "main": {
        "temp": 285.514,
        "pressure": 1013.75,
        "humidity": 100,
        "temp_min": 285.514,
        "temp_max": 285.514,
        "sea_level": 1023.22,
        "grnd_level": 1013.75
    },
    "wind": {
        "speed": 5.52,
        "deg": 311
    },
    "clouds": {
        "all": 0
    },
    "dt": 1485792967,
    "sys": {
        "message": 0.0025,
        "country": "JP",
        "sunris

In [24]:
with open('../APIs/openweathermap.json') as json_file:
  raw_json = json.load(json_file)

openweathermap_df = json_normalize(raw_json) 
openweathermap_df

base  clouds.all  cod  coord.lat  coord.lon          dt       id  \
0  stations           0  200      35.02     139.01  1485792967  1907296   

   main.grnd_level  main.humidity  main.pressure    ...      main.temp_max  \
0          1013.75            100        1013.75    ...            285.514   

   main.temp_min      name  sys.country sys.message sys.sunrise  sys.sunset  \
0        285.514  Tawarano           JP      0.0025  1485726240  1485763863   

                                             weather  wind.deg wind.speed  
0  [{'id': 800, 'main': 'Clear', 'description': '...       311       5.52  

[1 rows x 22 columns]

In [25]:
openweathermap_df.dtypes

base                object
clouds.all           int64
cod                  int64
coord.lat          float64
coord.lon          float64
dt                   int64
id                   int64
main.grnd_level    float64
main.humidity        int64
main.pressure      float64
main.sea_level     float64
main.temp          float64
main.temp_max      float64
main.temp_min      float64
name                object
sys.country         object
sys.message        float64
sys.sunrise          int64
sys.sunset           int64
weather             object
wind.deg             int64
wind.speed         float64
dtype: object

_May of the data types are already numeric.  Pretty cool. Let's dive into the weather object._

In [26]:
with open('../APIs/openweathermap.json') as json_file:
  raw_json = json.load(json_file)

openweathermap_weather_df = json_normalize(raw_json['weather']) 
openweathermap_weather_df

description icon   id   main
0   clear sky  01n  800  Clear

# TO-DO: Splice these normalized columns (aboe) back into main table

# One folder, one dataframe from multiple json files

The yahoo\_weather sufolder contains two json files from the Yahoo weather API.  We've moved all the linefeeds, but that won't affect our ability to make dataframes

In [27]:
! ls -1 ../APIs/yahoo_weather

yahoo_weather1line.json
yahoo_weather2.json


_Uncomment the lines below to see the contets of the files_

In [28]:
#! head ../APIs/yahoo_weather/yahoo_weather1line.json

#! echo ""
#! echo "===== NEXT FILE ==="
#! echo ""

#! cat ../APIs/yahoo_weather/yahoo_weather2.json

In [29]:
top_folder = "../APIs/yahoo_weather"
first_time = True

for filename in os.listdir(top_folder):
    filepath = top_folder + "/" + filename

    with open(filepath) as json_file:
      raw_json = json.load(json_file)

    new_json = json_normalize(raw_json['query']['results']['channel']['atmosphere'])
    
    if (first_time):
        first_time = False
        multifile_df = new_json
    else:
        multifile_df = multifile_df.append(new_json, ignore_index=True)
    
multifile_df

humidity pressure rising visibility
0       44   1019.0      0       16.1
1       58   1006.0      0       16.1

# Testing Wenfan's weather json, loaded from bitnami

In [30]:
# !ls /home/vagrant/from_bitnami/wenfan.old
# !head /home/vagrant/from_bitnami/wenfan.old
df = pd.read_json('/home/vagrant/from_bitnami/wenfan.old/2017-08-13-22-38-57')
df

BarometricPressure   Latitude   Longitude  PrecipitationInInches  \
0                1010.6  47.474800 -122.270400                    NaN   
1                1014.8  47.760633 -122.184048                    NaN   
2                 981.5  47.509000 -121.885000                    NaN   
3                1002.9  47.726000 -122.324000                    NaN   
4                   NaN  46.436000 -117.350000                    NaN   
5                   NaN  47.033800 -123.097100                    NaN   
6                   NaN  47.555500 -120.586600                    NaN   
7                   NaN  47.269100 -121.239200                    NaN   
8                   NaN  48.070300 -124.049000                    NaN   
9                   NaN  46.887000 -120.427000                    NaN   
10                  NaN  46.928000 -120.500000                    NaN   
11                  NaN  47.305082 -121.305688                    NaN   
12                  NaN  47.078000 -120.724000                    NaN   
13                  NaN  46.943000 -120.165000                    NaN   
14                996.5  48.688000 -122.400000                    NaN   
15                  NaN  47.745900 -121.089100                    NaN   
16               1013.7  48.992000 -122.272000                    NaN   
17                  NaN  47.043000 -119.916000                    NaN   
18                  NaN  47.085000 -119.006000                    NaN   
19                  NaN  47.113000 -120.786000                    NaN   
20                  NaN  48.050400 -122.952300                    NaN   
21                  NaN  48.093800 -123.278300                    NaN   
22                  NaN  47.443500 -121.669400                    NaN   
23                995.1  47.473000 -121.742000                    NaN   
24                  NaN  47.427000 -121.418000                    NaN   
25                  NaN  47.334800 -120.577100                    NaN   
26                  NaN  46.638000 -121.390000                    NaN   
27                  NaN  47.301900 -120.067800                    NaN   
28                  NaN  47.233100 -119.958200                    NaN   
29                  NaN  47.815600 -119.363000                    NaN   
..                  ...        ...         ...                    ...   
73                  NaN  48.876000 -117.348300                    NaN   
74                998.8  47.532000 -121.985000                    NaN   
75                987.6  46.217600 -119.748900                    NaN   
76                976.8  46.538500 -120.471800                    NaN   
77                977.7  46.538400 -120.472000                    NaN   
78                  NaN  46.393000 -120.241100                    NaN   
79                981.9  46.326600 -120.077900                    NaN   
80                  NaN  47.067700 -122.715800                    NaN   
81                874.5  48.390400 -119.879200                    NaN   
82                  NaN  47.858400 -122.621600                    NaN   
83               1015.5  48.446000 -122.432000                    NaN   
84               1000.2  47.916300 -122.207200                    NaN   
85                926.8  47.980300 -122.188000                    NaN   
86               1012.0  48.204200 -122.214200                    NaN   
87                996.5  47.813900 -121.570900                    NaN   
88                882.0  46.500278 -119.611667                    NaN   
89                  NaN  46.467063 -119.541111                    NaN   
90                  NaN  46.383155 -119.422786                    NaN   
91                  NaN  46.325574 -119.316126                    NaN   
92               1011.2  47.303516 -122.232946                    NaN   
93                605.5  46.049483 -118.495917                    NaN   
94               1003.0  47.830967 -122.262359                    NaN   
95                  NaN  46.802261 -123.007164                 

In [31]:
df.dtypes

BarometricPressure         float64
Latitude                   float64
Longitude                  float64
PrecipitationInInches      float64
ReadingTime                 object
RelativeHumidity           float64
SkyCoverage                 object
StationID                    int64
StationName                 object
TemperatureInFahrenheit    float64
Visibility                 float64
WindDirection              float64
WindDirectionCardinal       object
WindGustSpeedInMPH         float64
WindSpeedInMPH             float64
dtype: object

In [32]:
top_folder = "/home/vagrant/from_bitnami/wenfan/weather"
first_time = True

for filename in os.listdir(top_folder):
    filepath = top_folder + "/" + filename

    with open(filepath) as json_file:
      raw_json = json.load(json_file)

    #new_json = json_normalize(raw_json['query']['results']['channel']['atmosphere'])
    new_json = json_normalize(raw_json)
    
    if (first_time):
        first_time = False
        multifile_df = new_json
    else:
        multifile_df = multifile_df.append(new_json, ignore_index=True)
    
multifile_df

BarometricPressure   Latitude   Longitude PrecipitationInInches  \
0                  1011.7  47.474800 -122.270400                  None   
1                  1016.9  47.760633 -122.184048                  None   
2                   983.9  47.509000 -121.885000                  None   
3                  1004.7  47.726000 -122.324000                  None   
4                     NaN  46.436000 -117.350000                  None   
5                     NaN  47.033800 -123.097100                  None   
6                     NaN  47.555500 -120.586600                  None   
7                     NaN  47.269100 -121.239200                  None   
8                     NaN  48.070300 -124.049000                  None   
9                     NaN  46.887000 -120.427000                  None   
10                    NaN  46.928000 -120.500000                  None   
11                    NaN  47.305082 -121.305688                  None   
12                    NaN  47.078000 -120.724000                  None   
13                    NaN  46.943000 -120.165000                  None   
14                  999.0  48.688000 -122.400000                  None   
15                    NaN  47.745900 -121.089100                  None   
16                 1016.5  48.992000 -122.272000                  None   
17                    NaN  47.043000 -119.916000                  None   
18                    NaN  47.085000 -119.006000                  None   
19                    NaN  47.113000 -120.786000                  None   
20                    NaN  48.050400 -122.952300                  None   
21                    NaN  48.093800 -123.278300                  None   
22                    NaN  45.569000 -122.203000                  None   
23                    NaN  47.443500 -121.669400                  None   
24                  997.6  47.473000 -121.742000                  None   
25                    NaN  47.427000 -121.418000                  None   
26                    NaN  47.334800 -120.577100                  None   
27                    NaN  46.638000 -121.390000                  None   
28                    NaN  47.301900 -120.067800                  None   
29                    NaN  47.233100 -119.958200                  None   
...                   ...        ...         ...                   ...   
16280                 NaN  48.876000 -117.348300                  None   
16281               995.9  47.532000 -121.985000                  None   
16282               986.3  46.217600 -119.748900                  None   
16283               975.2  46.538500 -120.471800                  None   
16284               976.1  46.538400 -120.472000                  None   
16285                 NaN  46.393000 -120.241100                  None   
16286               980.1  46.326600 -120.077900                  None   
16287                 NaN  47.067700 -122.715800                  None   
16288               874.8  48.390400 -119.879200                  None   
16289                 NaN  47.858400 -122.621600                  None   
16290              1012.7  48.446000 -122.432000                  None   
16291               996.9  47.916300 -122.207200                  None   
16292               903.3  47.980300 -122.188000                  None   
16293              1008.4  48.204200 -122.214200                  None   
16294               993.1  47.813900 -121.570900                  None   
16295               882.0  46.500278 -119.611667                  None   
16296                 NaN  46.467063 -119.541111                  None   
16297                 NaN  46.383155 -119.422786                  None   
16298                 NaN  46.325574 -119.316126                  None   
16299              1008.6  47.303516 -122.232946                  None   
16300               605.5  46.049483 -118.495917                  None   
16301               999.8  47.830967 -122.262359                  None   
16302      

In [33]:
multifile_df.dtypes

BarometricPressure         float64
Latitude                   float64
Longitude                  float64
PrecipitationInInches       object
ReadingTime                 object
RelativeHumidity           float64
SkyCoverage                 object
StationID                    int64
StationName                 object
TemperatureInFahrenheit    float64
Visibility                  object
WindDirection              float64
WindDirectionCardinal       object
WindGustSpeedInMPH         float64
WindSpeedInMPH             float64
dtype: object

# One file, multiple json objects

_File 2yahoo.json contains two json objects from the yahoo weather api.  Each object has had newlines removed, so it only occupies one line of text. Uncomment the line below to have a look._

In [34]:
# !cat ../APIs/2yahoo.json

_The linux word count (wc) with the line (-l) option can count the lines int he file)_

In [35]:
!wc -l < ../APIs/2yahoo.json

2


_Read the json file into a dataframe ... what do we get?_

In [36]:
df3 = pd.read_json('../APIs/2yahoo.json', lines=True)
df3

query
0  {'count': 1, 'created': '2017-07-29T23:51:25Z'...
1  {'count': 1, 'created': '2017-08-01T02:14:07Z'...

_... not much.  Nice that we got a mulit-row dataframe, but this json needs to be normalized to be useful_

Code is pretty similar to the multi-file case, just using json.loads() instead of json.load()
Source: [Stack Overflow](https://stackoverflow.com/questions/8009882/how-to-read-large-file-line-by-line-in-python)

In [37]:
first_time = True

with open('../APIs/2yahoo.json') as f_one_json_per_line:
    for line in f_one_json_per_line:
        raw_json = json.loads(line)
        new_json = json_normalize(raw_json['query']['results']['channel']['atmosphere'])
        
        if (first_time):
            first_time = False
            multiline_df = new_json
        else:
            multiline_df = multiline_df.append(new_json, ignore_index=True)
        
multiline_df

humidity pressure rising visibility
0       58   1006.0      0       16.1
1       44   1019.0      0       16.1

# Testing Shiva's weather data, loaded from bitnami

In [38]:
!ls ../../../../../../from_bitnami/shiva
!wc -l < ../../../../../../from_bitnami/shiva/weather.json
!wc -l < ../../../../../../from_bitnami/json/2017-08-13-22-38-57
!wc -l < weather_yahoo_producer.py

weather2.json  weather.json
0
/bin/sh: 1: cannot open ../../../../../../from_bitnami/json/2017-08-13-22-38-57: No such file
60


In [39]:
!head ../../../../../../from_bitnami/shiva/weather.json

'clouds': {u'all': 1}, u'name': u'Mountain View', u'visibility': 16093, u'sys': {u'country': u'US', u'sunset': 1485826300, u'message': 0.0116, u'type': 1, u'id': 471, u'sunrise': 1485789140}, u'weather': [{u'main': u'Rain', u'id': 500, u'icon': u'10n', u'description': u'light rain'}], u'coord': {u'lat': 37.39, u'lon': -122.08}, u'base': u'stations', u'dt': 1485788160, u'main': {u'pressure': 1025, u'temp_min': 275.15, u'temp_max': 279.15, u'temp': 277.14, u'humidity': 86}, u'id': 5375480, u'wind': {u'speed': 1.67, u'deg': 53.0005}, u'cod': 200'clouds': {u'all': 1}, u'name': u'Mountain View', u'visibility': 16093, u'sys': {u'country': u'US', u'sunset': 1485826300, u'message': 0.0116, u'type': 1, u'id': 471, u'sunrise': 1485789140}, u'weather': [{u'main': u'Rain', u'id': 500, u'icon': u'10n', u'description': u'light rain'}], u'coord': {u'lat': 37.39, u'lon': -122.08}, u'base': u'stations', u'dt': 1485788160, u'main': {u'pressure': 1025, u'temp_min': 275.15, u'temp_max': 279.15, u'temp': 2

In [40]:
# ** RAISES ERRORS **
# df = pd.read_json('../../../../../../from_bitnami/shiva/weather2.json')
#df
#df.dtypes

# Exploring WSDOT Weather Data

In [41]:
# wsdot_Weather_df = pd.read_json('../APIs/wsdot.Traveltimes.json')
# wsdot_TravelTimes_df

# Exploring WSDOT Transportation Data

_Begin by loading the WSDOT Transportaiton Traveltims json file_

In [42]:
wsdot_TravelTimes_df = pd.read_json('../APIs/wsdot.Traveltimes.json')
wsdot_TravelTimes_df.head()

AverageTime  CurrentTime                                    Description  \
0           30           33    Everett to Downtown Seattle using HOV lanes   
1           30           29    Downtown Seattle to Everett using HOV lanes   
2           30           29                    Downtown Seattle to Everett   
3           11           11                  Downtown Bellevue to Issaquah   
4           11           11  Downtown Bellevue to Issaquah using HOV lanes   

   Distance                                           EndPoint  \
0     26.72  {'Description': 'I-5 @ University St in Seattl...   
1     26.94  {'Description': 'I-5 @ 41st St in Everett', 'D...   
2     26.94  {'Description': 'I-5 @ 41st St in Everett', 'D...   
3      9.55  {'Description': 'I-90 @ Front St in Issaquah',...   
4      9.55  {'Description': 'I-90 @ Front St in Issaquah',...   

                            Name  \
0    HOV Everett-Seattle (SB AM)   
1    HOV Seattle-Everett (NB PM)   
2        Seattle-Everett (NB PM)   
3      Bellevue-Issaquah (EB PM)   
4  HOV Bellevue-Issaquah (EB PM)   

                                          StartPoint  \
0  {'Description': 'I-5 @ 41st St in Everett', 'D...   
1  {'Description': 'I-5 @ University St in Seattl...   
2  {'Description': 'I-5 @ University St in Seattl...   
3  {'Description': 'I-405 @ NE 8th St in Bellevue...   
4  {'Description': 'I-405 @ NE 8th St in Bellevue...   

                  TimeUpdated  TravelTimeID  
0  /Date(1500575100000-0700)/             2  
1  /Date(1500575100000-0700)/             3  
2  /Date(1500575100000-0700)/             4  
3  /Date(1500575100000-0700)/             5  
4  /Date(1500575100000-0700)/             6

In [43]:
wsdot_TravelTimes_df.dtypes

AverageTime       int64
CurrentTime       int64
Description      object
Distance        float64
EndPoint         object
Name             object
StartPoint       object
TimeUpdated      object
TravelTimeID      int64
dtype: object

In [44]:
wsdot_TravelTimes_df.describe()

AverageTime  CurrentTime    Distance  TravelTimeID
count   180.000000   180.000000  180.000000    180.000000
mean     17.833333    19.316667   14.979167    182.277778
std      11.367158    12.569127    9.956458    145.362197
min       0.000000     0.000000    0.000000      1.000000
25%      10.000000    11.000000    9.410000     45.750000
50%      16.000000    16.000000   13.550000    118.000000
75%      23.000000    26.000000   17.927500    340.250000
max      78.000000    74.000000   78.560000    403.000000

_Fairly flat structure; some of the top-lvel types are numeric already. Let's try to crack open the STartpoint and Endpoint, so we can get some lat/long data._  *Note: this code will raise an error

In [45]:
# with open('../APIs/wsdot.Traveltimes.json') as json_file:
#  raw_json = json.load(json_file)

# wsdot_TravelTimes_df = json_normalize(raw_json['StartPoint']) 
# wsdot_TravelTimes_df

_Turns out we need to use row-indexing (below), but that only gets us one row at time._ 

In [46]:
with open('../APIs/wsdot.Traveltimes.json') as json_file:
  raw_json = json.load(json_file)

wsdot_TravelTimes_df = json_normalize(raw_json[0]['StartPoint'])
print(wsdot_TravelTimes_df)

                Description Direction  Latitude  Longitude  MilePost RoadName
0  I-5 @ 41st St in Everett         S  47.92428 -122.26548    192.55      I-5


In [47]:
wsdot_TravelTimes_df.dtypes

Description     object
Direction       object
Latitude       float64
Longitude      float64
MilePost       float64
RoadName        object
dtype: object

To normalize everything, we need to ....

1. Load the json file as a list of lines, so we can iterate over it.

2. Also load the json file as a raw json object.

3. Iterate over the lines, appendeing each one to a "master" dataframe

In [48]:
# Load the json as a list, so we can iterate over it.
with open('../APIs/wsdot.Traveltimes.json') as json_file:
  file_data = json_file.read()
    
list_json = json.loads(file_data)


# Load the json file as json, so we can normalize it.
with open('../APIs/wsdot.Traveltimes.json') as json_file:
  raw_json = json.load(json_file)


#iterate over the list_json. for each row, normalize the row_json
ii = 0

for travelTime in list_json:
    travel_df = json_normalize(raw_json[ii]['StartPoint'])
    if (ii == 0):
        wsdot_TravelTimesStartPoint_df= json_normalize(raw_json[ii]['StartPoint'])
    else:
        wsdot_TravelTimesStartPoint_df = wsdot_TravelTimesStartPoint_df.append(travel_df, ignore_index=True)
    ii= ii+ 1
   
wsdot_TravelTimesStartPoint_df

Description Direction   Latitude  \
0                            I-5 @ 41st St in Everett         S  47.924280   
1                      I-5 @ University St in Seattle         N  47.609294   
2                      I-5 @ University St in Seattle         N  47.609294   
3                       I-405 @ NE 8th St in Bellevue         S  47.613800   
4                       I-405 @ NE 8th St in Bellevue         S  47.613800   
5                         I-90 @ Front St in Issaquah         W  47.541799   
6                         I-90 @ Front St in Issaquah         W  47.541799   
7                       I-405 @ NE 8th St in Bellevue         N  47.613610   
8                            I-5 @ 41st St in Everett         S  47.924280   
9                       I-405 @ NE 8th St in Bellevue         N  47.613610   
10                           I-5 @ 41st St in Everett         S  47.924280   
11                    I-5 @ S 320th St in Federal Way         N  47.315446   
12                    I-5 @ S 320th St in Federal Way         N  47.315446   
13                         I-5 @ S 188th St in SeaTac         S  47.438760   
14                         I-5 @ S 188th St in SeaTac         S  47.438760   
15                    I-5 @ S 320th St in Federal Way         N  47.315446   
16                    I-5 @ S 320th St in Federal Way         N  47.315446   
17                     I-5 @ University St in Seattle         S  47.609294   
18                     I-5 @ University St in Seattle         S  47.609294   
19                      I-405 @ NE 8th St in Bellevue         S  47.613800   
20                    I-5 @ S 320th St in Federal Way         N  47.315446   
21                      I-405 @ NE 8th St in Bellevue         S  47.613800   
22                    I-5 @ S 320th St in Federal Way         N  47.315446   
23                     I-5 @ University St in Seattle         N  47.609294   
24                      I-5 @ 196th St SW in Lynnwood         S  47.817269   
25                     I-5 @ University St in Seattle         N  47.609294   
26                      I-5 @ 196th St SW in Lynnwood         S  47.817269   
27                      I-5 @ 196th St SW in Lynnwood         S  47.817269   
28                           I-405 @ SR 167 in Renton         S  47.467434   
29                     I-5 @ University St in Seattle         S  47.609294   
..                                                ...       ...        ...   
150                        SR14 and Union St in Camas         W  45.580300   
151                  I-5 SB and 179th St in Vancouver         S  45.754301   
152                     I-5 and 179th St in Vancouver         S  45.754301   
153                     I-5 and 179th St in Vancouver         S  45.754301   
154                            Delta Park in Portland         N  45.611460   
155                        Fremont Bridge in Portland         N  45.547310   
156                            I-405/US26 in Portland         N  45.516880   
157                                  I-84 in Portland         N  45.534880   
158               Portland International Airport Exit         N  45.572100   
159                          Rose Quarter in Portland         N  45.536250   
160                          Rose Quarter in Portland         E  45.520480   
161                          Rose Quarter in Portland         N  45.536250   
162                                       Tualatin OR         N  45.372970   
163                           Tualatin OR in Portland         N  45.353550   
164                         I-5 and 63rd in Vancouver         S  45.668141   
165                      I-5 SB and 63rd in Vancouver         S  45.668141   
166                              I-5 and Fourth Plain         N  45.636340   
167  I-205 SB and 18th St to SR14 to I-5 in Vancouver         S  45.628406   
168                 I-205 SB and 18th St in Vancouver         S  45.640957   
169                   I-5 @ Federal Way VMS MP 145.22         S 

In [49]:
wsdot_TravelTimesStartPoint_df.describe()

Latitude   Longitude    MilePost
count  180.000000  180.000000  180.000000
mean    46.301362 -120.312675   88.487111
std      6.102723   15.708814   80.749049
min      0.000000 -122.891484    0.100000
25%     46.672680 -122.508103   13.330000
50%     47.464051 -122.331130   71.025000
75%     47.613610 -122.206552  165.830000
max     48.152323    0.000000  307.900000

In [50]:
# Load the json as a list, so we can iterate over it.
with open('../APIs/wsdot.Traveltimes.json') as json_file:
  file_data = json_file.read()
    
list_json = json.loads(file_data)


# Load the json file as json, so we can normalize it.
with open('../APIs/wsdot.Traveltimes.json') as json_file:
  raw_json = json.load(json_file)


#iterate over the list_json. for each row, normalize the row_json
ii = 0

for travelTime in list_json:
  travel_df = json_normalize(raw_json[ii]['EndPoint'])
  if (ii == 0):
    wsdot_TravelTimesEndPoint_df= json_normalize(raw_json[ii]['EndPoint'])
  else:
    wsdot_TravelTimesEndPoint_df = wsdot_TravelTimesEndPoint_df.append(travel_df, ignore_index=True)
  ii= ii+ 1
   
wsdot_TravelTimesEndPoint_df

Description Direction   Latitude   Longitude  \
0             I-5 @ University St in Seattle         S  47.609294 -122.331759   
1                   I-5 @ 41st St in Everett         N  47.924280 -122.265480   
2                   I-5 @ 41st St in Everett         N  47.924280 -122.265480   
3                I-90 @ Front St in Issaquah         E  47.541799 -122.037396   
4                I-90 @ Front St in Issaquah         E  47.541799 -122.037396   
5              I-405 @ NE 8th St in Bellevue         N  47.613610 -122.187970   
6              I-405 @ NE 8th St in Bellevue         N  47.613610 -122.187970   
7                   I-5 @ 41st St in Everett         N  47.924280 -122.265480   
8              I-405 @ NE 8th St in Bellevue         S  47.613800 -122.188920   
9                   I-5 @ 41st St in Everett         N  47.924280 -122.265480   
10             I-405 @ NE 8th St in Bellevue         S  47.613800 -122.188920   
11                I-5 @ S 188th St in SeaTac         N  47.438760 -122.269470   
12                I-5 @ S 188th St in SeaTac         N  47.438760 -122.269470   
13           I-5 @ S 320th St in Federal Way         S  47.313840 -122.298609   
14           I-5 @ S 320th St in Federal Way         S  47.313840 -122.298609   
15            I-5 @ University St in Seattle         N  47.609294 -122.331759   
16            I-5 @ University St in Seattle         N  47.609294 -122.331759   
17           I-5 @ S 320th St in Federal Way         S  47.313840 -122.298609   
18           I-5 @ S 320th St in Federal Way         S  47.313840 -122.298609   
19           I-5 @ S 320th St in Federal Way         S  47.313840 -122.298609   
20             I-405 @ NE 8th St in Bellevue         N  47.613610 -122.187970   
21           I-5 @ S 320th St in Federal Way         S  47.313840 -122.298609   
22             I-405 @ NE 8th St in Bellevue         N  47.613610 -122.187970   
23             I-5 @ 196th St SW in Lynnwood         N  47.819545 -122.279109   
24            I-5 @ University St in Seattle         S  47.609294 -122.331759   
25             I-5 @ 196th St SW in Lynnwood         N  47.819545 -122.279109   
26            I-5 @ University St in Seattle         N  47.609294 -122.331759   
27            I-5 @ University St in Seattle         N  47.609294 -122.331759   
28            I-5 @ University St in Seattle         N  47.609294 -122.331759   
29                  I-405 @ SR 167 in Renton         N  47.467434 -122.219631   
..                                       ...       ...        ...         ...   
150             I-5 and I-84 JCT in Portland         W  45.525600 -122.660460   
151               I-5 Delta Park in Portland         S  45.603280 -122.683520   
152        I-205 and Airport Way in Portland         S  45.571880 -122.550940   
153         I-5 and Rose Quarter in Portland         S  45.536250 -122.667270   
154            I-5 and 179th St in Vancouver         N  45.754301 -122.663841   
155  I-5 and 78th St Hazel Dell in Vancouver         N  45.676544 -122.664559   
156  I-5 and 78th St Hazel Dell in Vancouver         N  45.676544 -122.664559   
157                  Mill Plain in Vancouver         N  45.621409 -122.559835   
158                 Clark County Fairgrounds         N  45.754301 -122.663841   
159                                    Camas         E  45.580300 -122.385324   
160                                    Camas         E  45.580300 -122.385324   
161            I-5 and 179th St in Vancouver         N  45.754301 -122.663841   
162                          Salmon Creek WA         N  45.707495 -122.631733   
163            I-5 and 134th St in Vancouver         N  45.716985 -122.654515   
164            I-5 and I-405 JCT in Portland         S  45.556320 -122.679010   
165                         US26 in Portland         S  45.513480 -122.667930   
166                           I-5 and SR 502         N  45.781266 -122.672017   
167             I-5 and I-84 JCT in Portland         S  45.54485

In [51]:
wsdot_TravelTimesEndPoint_df.describe()

Latitude   Longitude    MilePost
count  180.000000  180.000000  180.000000
mean    47.072796 -122.366538   99.269722
std      0.859379    0.241429   89.226318
min     45.353550 -122.891484    0.110000
25%     47.047885 -122.560915   13.330000
50%     47.464051 -122.331759  116.520000
75%     47.613610 -122.219631  165.830000
max     48.152323 -120.599140  307.350000

# Joining Transportation and Weather with geopandas
- Gather up the data frames (one weather, one transporattion, both in same proximity)
- Load the geopandas library
- datafraems to geoframes
- do the join

_We have a travel dataframe from WSDOT. It has "Latitude" and "Longituded" cordinate fields_

In [52]:
wsdot_TravelTimesEndPoint_df.head()

Description Direction   Latitude   Longitude  MilePost  \
0  I-5 @ University St in Seattle         S  47.609294 -122.331759    165.83   
1        I-5 @ 41st St in Everett         N  47.924280 -122.265480    192.77   
2        I-5 @ 41st St in Everett         N  47.924280 -122.265480    192.77   
3     I-90 @ Front St in Issaquah         E  47.541799 -122.037396     16.96   
4     I-90 @ Front St in Issaquah         E  47.541799 -122.037396     16.96   

  RoadName  
0      I-5  
1      I-5  
2      I-5  
3     I-90  
4     I-90

_We also have an OpenWeatherMap dataframe. It has "coord.lat" and "coord.lon"_

In [53]:
with open('../APIs/openweathermap.json') as json_file:
  raw_json = json.load(json_file)

openweathermap_df = json_normalize(raw_json) 
openweathermap_df

base  clouds.all  cod  coord.lat  coord.lon          dt       id  \
0  stations           0  200      35.02     139.01  1485792967  1907296   

   main.grnd_level  main.humidity  main.pressure    ...      main.temp_max  \
0          1013.75            100        1013.75    ...            285.514   

   main.temp_min      name  sys.country sys.message sys.sunrise  sys.sunset  \
0        285.514  Tawarano           JP      0.0025  1485726240  1485763863   

                                             weather  wind.deg wind.speed  
0  [{'id': 800, 'main': 'Clear', 'description': '...       311       5.52  

[1 rows x 22 columns]

# Here's what finally worked

from http://geoffboeing.com/2015/10/exporting-python-data-geojson/

In [54]:
def df_to_geojson(df, properties, lat='latitude', lon='longitude'):
    geojson = {'type':'FeatureCollection', 'features':[]}
    for _, row in df.iterrows():
        feature = {'type':'Feature',
                   'properties':{},
                   'geometry':{'type':'Point',
                               'coordinates':[]}}
        feature['geometry']['coordinates'] = [row[lon],row[lat]]
        for prop in properties:
            feature['properties'][prop] = row[prop]
        geojson['features'].append(feature)
    return geojson

In [55]:
openweathermap_df.dtypes

base                object
clouds.all           int64
cod                  int64
coord.lat          float64
coord.lon          float64
dt                   int64
id                   int64
main.grnd_level    float64
main.humidity        int64
main.pressure      float64
main.sea_level     float64
main.temp          float64
main.temp_max      float64
main.temp_min      float64
name                object
sys.country         object
sys.message        float64
sys.sunrise          int64
sys.sunset           int64
weather             object
wind.deg             int64
wind.speed         float64
dtype: object

In [56]:
cols = ['base','clouds.all', 'cod','dt','id','main.grnd_level','main.humidity',
        'main.pressure','main.sea_level','main.temp','main.temp_max','main.temp_min',
        'name','sys.country','sys.message','sys.sunrise','sys.sunset','weather','wind.deg','wind.speed']
        
geojson = df_to_geojson(openweathermap_df, cols, lat='coord.lat', lon='coord.lon')

In [57]:
geoWeather = geopandas.GeoDataFrame.from_features(geojson['features'])

In [58]:
geoWeather

base  clouds.all  cod          dt              geometry       id  \
0  stations           0  200  1485792967  POINT (139.01 35.02)  1907296   

   main.grnd_level  main.humidity  main.pressure  main.sea_level     ...      \
0          1013.75            100        1013.75         1023.22     ...       

   main.temp_max  main.temp_min      name sys.country sys.message  \
0        285.514        285.514  Tawarano          JP      0.0025   

   sys.sunrise  sys.sunset                                            weather  \
0   1485726240  1485763863  [{'id': 800, 'main': 'Clear', 'description': '...   

  wind.deg  wind.speed  
0      311        5.52  

[1 rows x 21 columns]

In [59]:
wsdot_TravelTimesStartPoint_df.dtypes

Description     object
Direction       object
Latitude       float64
Longitude      float64
MilePost       float64
RoadName        object
dtype: object

In [60]:
cols = ['Description','Direction', 'MilePost', 'RoadName']
geojson = df_to_geojson(wsdot_TravelTimesStartPoint_df, cols, lat='Latitude', lon='Longitude')
geoTraffic = geopandas.GeoDataFrame.from_features(geojson['features'])

In [61]:
geoWeather.head()

base  clouds.all  cod          dt              geometry       id  \
0  stations           0  200  1485792967  POINT (139.01 35.02)  1907296   

   main.grnd_level  main.humidity  main.pressure  main.sea_level     ...      \
0          1013.75            100        1013.75         1023.22     ...       

   main.temp_max  main.temp_min      name sys.country sys.message  \
0        285.514        285.514  Tawarano          JP      0.0025   

   sys.sunrise  sys.sunset                                            weather  \
0   1485726240  1485763863  [{'id': 800, 'main': 'Clear', 'description': '...   

  wind.deg  wind.speed  
0      311        5.52  

[1 rows x 21 columns]

In [62]:
geoTraffic.head()

Description Direction  MilePost RoadName  \
0        I-5 @ 41st St in Everett         S    192.55      I-5   
1  I-5 @ University St in Seattle         N    165.83      I-5   
2  I-5 @ University St in Seattle         N    165.83      I-5   
3   I-405 @ NE 8th St in Bellevue         S     13.33    I-405   
4   I-405 @ NE 8th St in Bellevue         S     13.33    I-405   

                        geometry  
0    POINT (-122.26548 47.92428)  
1  POINT (-122.331759 47.609294)  
2  POINT (-122.331759 47.609294)  
3     POINT (-122.18892 47.6138)  
4     POINT (-122.18892 47.6138)

# Mapping with Bokeh [under construction]

Source: http://bokeh.pydata.org/en/latest/docs/user_guide/geo.html

This shows promise http://michael-harmon.com/blog/IntroToBokeh.html
(48.152323,-122.891484)	(48.152323, -120.599140)
(45.353550,-122.891484)	(45.353550, -120.599140)

Other promising sources
Vincent maps: http://wrobstory.github.io/2013/04/python-maps-choropleth.html
Folium https://github.com/python-visualization/folium
Bog post, seems to be al matplotlib: http://www.datadependence.com/2016/06/creating-map-visualisations-in-python/
stackoverlfow: https://stackoverflow.com/questions/9018607/library-for-map-visualization


Source:http://michael-harmon.com/blog/IntroToBokeh.html

# BEGIN ==== WORKING =====

In [65]:
# source http://michael-harmon.com/blog/IntroToBokeh.html
from bokeh.io import output_file, output_notebook, show
from bokeh.plotting import figure, ColumnDataSource
output_notebook()

def convert_GeoPandas_to_Bokeh_format(gdf):
    """
    Function to convert a GeoPandas GeoDataFrame to a Bokeh
    ColumnDataSource object.
    
    :param: (GeoDataFrame) gdf: GeoPandas GeoDataFrame with polygon(s) under
                                the column name 'geometry.'
                                
    :return: ColumnDataSource for Bokeh.
    """
    gdf_new = gdf.drop('geometry', axis=1).copy()
    
    gdf_new['x'] = gdf.apply(getGeometryCoords, 
                             geom='geometry', 
                             coord_type='x', 
                             shape_type='point', 
                             axis=1)
    
    gdf_new['y'] = gdf.apply(getGeometryCoords, 
                             geom='geometry', 
                             coord_type='y', 
                             shape_type='point', 
                             axis=1)
    
    
    return ColumnDataSource(gdf_new)


def getGeometryCoords(row, geom, coord_type, shape_type):
    """
    Returns the coordinates ('x' or 'y') of edges of a Polygon exterior.
    
    :param: (GeoPandas Series) row : The row of each of the GeoPandas DataFrame.
    :param: (str) geom : The column name.
    :param: (str) coord_type : Whether it's 'x' or 'y' coordinate.
    :param: (str) shape_type
    """
    
    # Parse the exterior of the coordinate
    if shape_type == 'polygon':
        exterior = row[geom].geoms[0].exterior
        if coord_type == 'x':
            # Get the x coordinates of the exterior
            return list( exterior.coords.xy[0] )    
        
        elif coord_type == 'y':
            # Get the y coordinates of the exterior
            return list( exterior.coords.xy[1] )

    elif shape_type == 'point':
        exterior = row[geom]
    
        if coord_type == 'x':
            # Get the x coordinates of the exterior
            return  exterior.coords.xy[0][0] 

        elif coord_type == 'y':
            # Get the y coordinates of the exterior
            return  exterior.coords.xy[1][0]

Loading BokehJS ...

In [66]:
SEA_Source = convert_GeoPandas_to_Bokeh_format(geoTraffic)

In [67]:
from bokeh.io import output_notebook, show
from bokeh.models import (
  GMapPlot, GMapOptions, ColumnDataSource, Circle, DataRange1d, PanTool, WheelZoomTool, BoxSelectTool
)

map_options = GMapOptions(lat=46.252, lng=-120.65, map_type="roadmap", zoom=7)

plot = GMapPlot(
    x_range=DataRange1d(), y_range=DataRange1d(), map_options=map_options
)
plot.title.text = "Pacific Northwest"

# For GMaps to function, Google requires you obtain and enable an API key:
#
#     https://developers.google.com/maps/documentation/javascript/get-api-key
#
# Replace the value below with your personal API key:
plot.api_key = " AIzaSyC7C1EXGMAaeJgox2slsYiKaGVa_TUPWDg "


circle = Circle(x="x", y="y", size=15, fill_color="blue", fill_alpha=0.8, line_color=None)
plot.add_glyph(SEA_Source, circle)

plot.add_tools(PanTool(), WheelZoomTool(), BoxSelectTool())
output_file("gmap_plot.html")
# output_notebook()
show(plot)

# Future Work
- Figure out how to join and merge stuff (geogson)
- Parse dates from objects into something actionable
- Go back to the tranporation df's, do I really need to load twice?


In [68]:
from bokeh.io import output_notebook, show
from bokeh.models import GeoJSONDataSource
from bokeh.plotting import figure
from bokeh.sampledata.sample_geojson import geojson

# geo_source = GeoJSONDataSource(geojson=geojson)

p = figure()
p.circle(x='x', y='y', alpha=0.9, source=SEA_Source)
show(p)

# END ==== WORKING =====